File: 015.NN_UIViewAnimationPractice/NN_UIViewAnimationPractice/UIViewAnimationPractice/Controllers/NNFrameAnimationController.m
//
//  NNFrameAnimationController.m
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2018/8/11.
//  Copyright © 2018年 刘朋坤. All rights reserved.
//

#import "NNFrameAnimationController.h"

@interface NNFrameAnimationController ()

@end

@implementation NNFrameAnimationController

- (NSArray *)buttonTitleArray {
    return @[@"Frame", @"Bounds", @"Center"];
}

- (void)buttonClick:(UIButton *)sender {
    switch (sender.tag) {
        case 0:
            [self changeFrame];
            break;
        case 1:
            [self changeBounds];
            break;
        case 2:
            [self changeCenter];
            break;
    }
}

- (void)changeFrame {
    CGRect originalFrame = self.nnView.headerImage.frame;
    CGRect changeFrame = CGRectMake(self.nnView.headerImage.frame.origin.x, self.nnView.headerImage.frame.origin.y - 120, 200, 80);
    [self changeOriginalRect:originalFrame changeRect:changeFrame];
}

- (void)changeBounds {
    CGRect originalBounds = self.nnView.headerImage.bounds;
    CGRect changeBounds = CGRectMake(0, 0, self.view.frame.size.width - 100, 100);
    [self changeOriginalRect:originalBounds changeRect:changeBounds];
}

- (void)changeCenter {
    CGPoint originalPoint = self.nnView.headerImage.center;
    CGPoint changePoint = CGPointMake(self.nnView.headerImage.center.x, self.nnView.headerImage.center.y - 160);
    [UIView animateWithDuration:0.5 animations:^{
        self.nnView.headerImage.center = changePoint;
    } completion:^(BOOL finished) {
        [UIView animateWithDuration:1 animations:^{
            self.nnView.headerImage.center = originalPoint;
        }];
    }];
}

- (void)changeOriginalRect:(CGRect)originalRect changeRect:(CGRect)changeRect {
    [UIView animateWithDuration:0.5 animations:^{
        self.nnView.headerImage.bounds =  changeRect;
    } completion:^(BOOL finished) {
        [UIView animateWithDuration:1 animations:^{
            self.nnView.headerImage.bounds = originalRect;
        }];
    }];
}

@end

File: 017.NN_UIViewAnimationPractice_swift/NN_UIViewAnimationPractice_swift/UIViewAnimationPractice/Controllers/NNFrameAnimationController.swift
//
//  NNFrameAnimationController.swift
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2019/1/5.
//  Copyright © 2019 刘朋坤. All rights reserved.
//

import UIKit

class NNFrameAnimationController: NNBaseViewController {
    
    // 按钮点击事件
    override func buttonClick(index: NSInteger) {
        switch index {
        case 0:
            changeFrame()
            break
        case 1:
            changeBounds()
            break
        case 2:
            changeCenter()
            break
        default:
            break
        }
    }
    
    // 改变Frame
    func changeFrame() {
        let originalFrame = self.nnView.headerImage.frame
        let changeFrame = CGRect(x: self.nnView.headerImage.frame.origin.x, y: self.nnView.headerImage.frame.origin.y - 120, width: 200, height: 80)
        changeOriginalRect(originalFrame: originalFrame, changeFrame: changeFrame)
    }
    
    // 改变Bounds
    func changeBounds() {
        let originalBounds = self.nnView.headerImage.bounds
        let changeBounds = CGRect(x: 0, y: 0, width: self.view.frame.size.width - 100, height: 100)
        changeOriginalRect(originalFrame: originalBounds, changeFrame: changeBounds)
    }
    
    // 改变Center
    func changeCenter() {
        let originalPoint = self.nnView.headerImage.center
        let changePoint = CGPoint(x: self.nnView.headerImage.center.x, y: self.nnView.headerImage.center.y - 160)
        UIView.animate(withDuration: 0.5, animations: {
            self.nnView.headerImage.center = changePoint
        }) { (finished) in
            UIView.animate(withDuration: 1, animations: {
                self.nnView.headerImage.center = originalPoint
            })
        }
    }
    
    // 封装的通用方法
    func changeOriginalRect(originalFrame:CGRect, changeFrame:CGRect) {
        UIView.animate(withDuration: 0.5, animations: {
            self.nnView.headerImage.bounds = changeFrame
        }) { (finished) in
            UIView.animate(withDuration: 1, animations: {
                self.nnView.headerImage.bounds = originalFrame
            })
        }
    }
    
    // 数组元素
    override func buttonTitleArrayMethod() -> NSArray {
        return ["Frame", "Bounds", "Center"]
    }
    
}
